kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/batch/KCBQThreadPoolExecutor.java

```diff
@@ -106,4 +106,18 @@ private static String createErrorString(Collection<Throwable> errors) {
                         .collect(Collectors.toList()));
     return String.join(", ", exceptionTypeStrings);
   }
+
+  /**
+   * Checks for BigQuery errors. No-op if there isn't any error.
+   *
+   * @throws BigQueryConnectException if there is any error when writing to BigQuery.
+   */
+  public void checkForErrors() throws BigQueryConnectException {
+    synchronized (encounteredErrors) {
```

Comment:
I haven't had a chance to look at how this datastructure is used throughout. But just wanted to understand why we need to synchronize here. Is that absolutely necessary?

Reply:
I think if we don't synchronize here, when we try to create error string later it might have already been cleared in `afterExecute`.

Reply:
Is it possible that a call to `put` causes writes to fail, then a call to `flush` causes `encounteredErrors` to get cleared (and throws an exception), and then the next call to `put` calls `checkForErrors` but since `encounteredErrors` has been cleared it doesn't cause an exception to be thrown?

Reply:
I cannot eliminate this possibility. And that's why I put the error check after submitting tasks to the executor, because the insertion happens immediately, and why the wait solution is preferable. But I think this would be a very rare case.

Reply:
I think if we know that the task is going to fail the next time that `put` is called, it might make sense to not clear `encounteredErrors` when it's checked in either `checkForErrors` or `awaitCurrentTasks`... is there a case when we'd ever want to clear that?

Reply:
what are the implications of not clearing in `awaitCurrentTasks`. Wouldn't we keep accumulating the same errors again and again? May be its not too bad.

Reply:
I don't think so... we can consider a couple possible cases here:
- A call to `put` occurs and a resulting write fails, then a second call to `put` occurs and the task throws an exception and fails
- A call to `put` occurs and a resulting write fails, then `flush` is called and the task throws an exception, causing the offsets to rewind, then a second call to `put` occurs and the task throws an exception and fails
In neither of these cases will we see duplicate errors.

The only case I can think of where we would see duplicate errors would be this one:
- A call to `put` occurs and a resulting write fails, then `flush` is called and the task throws an exception, then _another_ call to `flush` occurs and the task throws the same exception all over again

This would be a pretty weird edge case, but even if it does happen, I think the behavior is somewhat reasonable. We wouldn't add new errors to `encounteredErrors` since that would only happen as the result of a call to `put`, which would throw an exception as soon as any of the writes fails with a non-retriable exception; we'd just end up throwing the same exception over and over again whenever `flush` got called until the framework finally invoked `put`.

Reply:
In scenario # 2 if we don't clear the error aren't we going to essentially get the same exception all over again? Its not a big deal but just wanted to call out the scenario.

Reply:
Yeah, you'll potentially see the same exception in stack traces coming from both `flush` and `put`. We could consider changing the exception that gets thrown in `flush` altogether so that instead of being specific it just says something like "Cannot flush (or maybe commit offsets) because writes to BigQuery have failed", and then throwing the specific exception from `put`. Might make slogging through log messages a little easier for new users, but might make things harder for existing users who expect logs to take a certain format.

I think it's probably fine as-is 🤷